src/core/vdom/patch.js

```diff
@@ -643,7 +644,12 @@ export function createPatchFunction (backend) {
             let childrenMatch = true
             let childNode = elm.firstChild
             for (let i = 0; i < children.length; i++) {
-              if (!childNode || !hydrate(childNode, children[i], insertedVnodeQueue, inVPre)) {
+              const child = children[i]
+              // ignore empty text vnode
+              if (isUndef(child.tag) && isFalse(child.isComment) && child.text === '') {
+                continue
```

Comment:
This results in child.elm === undefined and cause a crash in next patch.
Reproduction: https://github.com/deqwin/vue/compare/fix_ssr_hydration...contribu:fix_ssr_hydration_crash

Reply:
Thanks for your suggestions. i will see how to modify.

Reply:
The fact is not that the vnode tree has an extra empty text node, but the DOM tree is missing a necessary empty text node.

Reply:
@posva I have modified the code to deal with this problem.

Reply:
@posva I have modified the code to deal with this problem and also updated the test.

Reply:
Thanks!
I am unsure if this causes repainting during hydration and if the existing hydration code allows it.
I think it shouldn't cause repainting if the existing hydration code disallows repainting.

Reply:
Yeah, it‘s worth exploring. However, it is necessary to keep the vnode structure and DOM structure consistent at all times. So I think it's ok to do this step here.

Reply:
Another solution is rendering empty text nodes as comment nodes in SSR like async placeholder or v-if="false".
I am unsure which solution is better.